extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java

```diff
@@ -242,18 +242,35 @@ public void transform(TransformationContext context) {
             BuildProducer<ValidationPhaseBuildItem.ValidationErrorBuildItem> errors) {
-        Type generatedBeanType = Type.create(GrpcDotNames.MUTINY_BEAN, org.jboss.jandex.Type.Kind.CLASS);
-        for (BeanInfo bean : validationPhase.getContext().beans().classBeans().withBeanType(BindableService.class)) {
-            if (!bean.getTypes().contains(generatedBeanType) && bean.getQualifiers().stream().map(AnnotationInstance::name)
-                    .noneMatch(GrpcDotNames.GRPC_SERVICE::equals)) {
-                errors.produce(new ValidationPhaseBuildItem.ValidationErrorBuildItem(
-                        new IllegalStateException(
-                                "A gRPC service bean must be annotated with @io.quarkus.grpc.GrpcService: " + bean)));
-            }
-            if (!bean.getScope().getDotName().equals(BuiltinScope.SINGLETON.getName())) {
-                errors.produce(new ValidationPhaseBuildItem.ValidationErrorBuildItem(
-                        new IllegalStateException("A gRPC service bean must have the javax.inject.Singleton scope: " + bean)));
+        Type mutinyBeanType = Type.create(GrpcDotNames.MUTINY_BEAN, org.jboss.jandex.Type.Kind.CLASS);
+        Type mutinyServiceType = Type.create(GrpcDotNames.MUTINY_SERVICE, org.jboss.jandex.Type.Kind.CLASS);
+        Type bindableServiceType = Type.create(GrpcDotNames.BINDABLE_SERVICE, org.jboss.jandex.Type.Kind.CLASS);
+        Predicate<Set<Type>> predicate = new Predicate<Set<Type>>() {
+            @Override
+            public boolean test(Set<Type> types) {
+                return types.contains(bindableServiceType) || types.contains(mutinyServiceType);
             }
+        };
+        for (BeanInfo bean : validationPhase.getContext().beans().classBeans().matchBeanTypes(predicate)) {
+            validateBindableService(bean, mutinyBeanType, errors);
+        }
+        // Validate the removed beans as well - detect beans not annotated with @GrpcService
+        for (BeanInfo bean : validationPhase.getContext().removedBeans().classBeans().matchBeanTypes(predicate)) {
+            validateBindableService(bean, mutinyBeanType, errors);
+        }
+    }
+
+    private void validateBindableService(BeanInfo bean, Type generatedBeanType,
+            BuildProducer<ValidationPhaseBuildItem.ValidationErrorBuildItem> errors) {
+        if (!bean.getTypes().contains(generatedBeanType) && bean.getQualifiers().stream().map(AnnotationInstance::name)
+                .noneMatch(GrpcDotNames.GRPC_SERVICE::equals)) {
+            errors.produce(new ValidationPhaseBuildItem.ValidationErrorBuildItem(
+                    new IllegalStateException(
+                            "A gRPC service bean must be annotated with @io.quarkus.grpc.GrpcService: " + bean)));
+        }
+        if (!bean.getScope().getDotName().equals(BuiltinScope.SINGLETON.getName())) {
+            errors.produce(new ValidationPhaseBuildItem.ValidationErrorBuildItem(
```

Comment:
As we now  only support having services with the `@GrpcService` annotation, I'm wondering if we shouldn't change the error messages here

Reply:
Well, you can still annotate your bean with `@Singleton`, e.g. `@Singleton @GrpcService class MyBean {}`...

Reply:
but if you have the error you either added a wrong scope-defining annotation or some stereotype that broght in such annotation, right? Without it it would all be good, right?

Reply:
Yes. That's why the message reads "has scope" and not "is annotated with"...